### ISSUE TITLE
[no-implied-eval] False positive for logical OR coalesced expressions

### ISSUE THREAD

**Issue by jas14** [typescript-eslint/typescript-eslint#3621]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-implied-eval": ["error"]
  }
}
```

```TS
const foo = () => { /* noop */ }
const bar = () => { /* noop */ }

setTimeout(
  foo || bar,
  500
);
```

**Expected Result**

No error.

**Actual Result**

```
  5:3  error  Implied eval. Consider passing a function  @typescript-eslint/no-implied-eval
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.28.2` |
| `@typescript-eslint/parser`        | `4.28.2` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.30.0` |
| `node`                             | `16.4.2` |

**Comment by bradzacher** [comment]
bug is that we only allow a subset of expressions.

https://github.com/typescript-eslint/typescript-eslint/blob/8c65d30a225a3b99e80326961d0cb0c8189b039c/packages/eslint-plugin/src/rules/no-implied-eval.ts#L102-L124

could probably expand this to just check the type of everything else.
